virttest/cpu.py

```diff
@@ -1258,3 +1261,132 @@ def get_qemu_best_cpu_model(params):
+    """
+    Get the best cpu model and features available for qemu
+    """
+    cache_file = "qemu_best_cpu_cache"
+    cache_path = os.path.join(data_dir.get_tmp_dir(), cache_file)
+    cpu_info = get_cpu_info_from_cache(cache_path)
+    if not cpu_info:
+        cpu_info = {}
+        cpu_model = get_qemu_best_cpu_model(params)
+        cpu_features = ""
+        cpu_info["model"] = cpu_model
+        cpu_info["features"] = cpu_features
+        store_cpu_info_to_cache(cache_path, cpu_info)
+    return cpu_info
+
+
+def get_cpu_info_from_virsh(params):
+    """
+    Try to get cpu model and features from virsh with 'host-model'
+    """
+    vm_arch = params["vm_arch_name"]
+    if vm_arch not in ['x86_64', 'i686']:
+        raise NotImplementedError("Arch '%s' is not supported" % vm_arch)
+    machine = params.get('machine_type')
+    name = uuid.uuid4().hex
+    cache_file = "%s_%s_libvirt_host_model_cache" % (vm_arch, machine)
+    cache_path = os.path.join(data_dir.get_tmp_dir(), cache_file)
+    cpu_info = get_cpu_info_from_cache(cache_path)
+    if cpu_info:
+        return cpu_info
+    try:
+        path.find_command("virsh")
+    except path.CmdNotFoundError:
+        logging.warning("Virsh executable not set or found on path")
+        return
+
+    xml = """
+            <domain type='kvm'>
+                <name>%s</name>
+                <memory>1</memory>
+                <os>
+                    <type arch='%s' machine='%s'>hvm</type>
+                </os>
+                <cpu mode='host-model'/>
+            </domain>
+          """ % (name, vm_arch, machine)
+    xml_file = os.path.join(data_dir.get_tmp_dir(), "temp_xml_for_cpu")
+    with open(xml_file, "w") as f:
+        f.write(xml)
+    try:
+        logging.info("Get cpu model and features from virsh")
+        virsh.define(xml_file)
+        virsh.start(name)
+        cpu_xml = libvirt_xml.VMXML.new_from_dumpxml(name).cpu
+        feature_list = cpu_xml.get_feature_list()
+        cpu_model = cpu_xml.model
+        cpu_features = {}
+        for i in range(0, len(feature_list)):
+            feature_name = cpu_xml.get_feature(i).get('name')
+            feature_policy = cpu_xml.get_feature(i).get('policy')
+            if feature_policy == "require":
+                feature_policy = "on"
+            elif feature_policy == "disable":
+                feature_policy = "off"
+            cpu_features[feature_name] = feature_policy
+    except Exception as err:
+        logging.error(err)
+        return
+    else:
+        cpu_info = {}
+        cpu_info["model"] = cpu_model
+        cpu_info["features"] = cpu_features
+        store_cpu_info_to_cache(cache_path, cpu_info)
+        return cpu_info
+    finally:
+        if virsh.is_alive(name):
+            virsh.destroy(name, ignore_status=True)
+        virsh.undefine(name, ignore_status=True)
+
+
+def combine_flags(flags_dict, flags_string):
+    """
+    Combine the flags in flags_dict and flags_string
```

Comment:
A bit suggestion about the parameter name,  how do you think about indicate the source of the flags, and the format of the flags can be explained with the doc:
```
def combine_flags(auto_cpu_flags, custom_cpu_flag):
```

Reply:
I guess the function itself doesn't care if the flags are auto generated or defined by user. It cares more about the param type. I think the variable names are more useful when the function is called other than defined. Thanks.

Reply:
It depends on the purpose of this function, whether the implemention of this function is available for other dict + string combination, not only cpu flags. IMO, it is for cpu flags only in cpu.py,  correct me if my understanding is wrong.

Reply:
I'm not quite sure about the dependency.  Yes, it's only for cpu flags now. But as to function definition has to indicate the param source, I doubt it.